File: comm/src/main/scala/coop/rchain/comm/discovery/NodeDiscovery.scala
package coop.rchain.comm.discovery

import cats.tagless._
import cats.Monad
import coop.rchain.comm.{NodeIdentifier, PeerNode}

import coop.rchain.shared.Log

@autoFunctorK
@autoSemigroupalK
@autoProductNK
trait NodeDiscovery[F[_]] {
  def discover: F[Unit]
  def peers: F[Seq[PeerNode]]
}

object NodeDiscovery extends NodeDiscoveryInstances {
  def apply[F[_]](implicit L: NodeDiscovery[F]): NodeDiscovery[F] = L
}

sealed abstract class NodeDiscoveryInstances {

  def kademlia[F[_]: Monad: KademliaStore: KademliaRPC: Log](id: NodeIdentifier): NodeDiscovery[F] =
    new NodeDiscovery[F] {
      def discover: F[Unit]       = KademliaNodeDiscovery.discover(id)
      def peers: F[Seq[PeerNode]] = KademliaNodeDiscovery.peers
    }
}

File: comm/src/rust/discovery/node_discovery.rs
// See comm/src/main/scala/coop/rchain/comm/discovery/NodeDiscovery.scala

use crate::rust::{
    errors::CommError,
    peer_node::{NodeIdentifier, PeerNode},
};

use super::{
    kademlia_node_discovery::KademliaNodeDiscovery, kademlia_rpc::KademliaRPC,
    kademlia_store::KademliaStore,
};

pub trait NodeDiscovery {
    fn discover(&self) -> Result<(), CommError>;

    fn peers(&self) -> Result<Vec<PeerNode>, CommError>;
}

pub fn kademlia<'a, T: KademliaRPC + Clone>(
    id: NodeIdentifier,
    kademlia_rpc: &'a T,
) -> KademliaNodeDiscovery<'a, T> {
    KademliaNodeDiscovery::new(KademliaStore::new(id, kademlia_rpc), kademlia_rpc)
}
